### ISSUE TITLE
Kokkos allow HIP compilers that aren't called hipcc

### ISSUE THREAD

**Issue by jjellio** [kokkos/kokkos#5453]
This is a request!
Kokkos (through Trilinos - or always?) - wants to enable a TPL of ROCM if the compiler is not `hipcc`.
https://github.com/kokkos/kokkos/blob/develop/cmake/kokkos_tpls.cmake#L42

This is kinda annoying, and I religiously patch my Trilinos pulls to `SET(ROCM_DEFAULT OFF).`

The builds all work fine with amdclang or crayclang directly ... as these compilers imply ROCM (do any modern HIP compilers not?).  I just manually merged the latest kokkos promotion (since Trilinos' PR tester is .... having issues) .. and I keep hoping with all the Cmake updates, that this nuisance  will go away.... but it doesn't.

Will this ultimately go away at some point? 

**Please also include the following items to support reproducing the bug**
1. anything but `hipcc`
2. Kokkos promotion in PR testing at Trilinos (3.7)
3. any AMD machine using HIP
4. cmake configure command
```
 "-DKokkos_ARCH_ZEN3:BOOL=ON" \
 "-DKokkos_ENABLE_SERIAL:BOOL=ON" \
 "-DKokkos_ENABLE_OPENMP:BOOL=OFF" \
 "-DKokkos_ENABLE_HIP:BOOL=ON" \
 "-DKokkos_ARCH_VEGA90A:BOOL=ON" \
```
6. output from cmake command
7. code needed to reproduce the bug
8. command line needed to reproduce the bug
7. please also attach the `KokkosCore_config.h` header file (generated during the build);
**Any additional info**
Please provide any additional context about the issue here.

**Comment by Rombur** [comment]
I've just tried to set `SET(ROCM_DEFAULT OFF)` with `amdclang` and Kokkos does not compile.
```
[  5%] Linking CXX executable KokkosCore_UnitTest_PushFinalizeHook                                                                                                                                                 
ld.lld: error: undefined symbol: __hipRegisterVar                                                                                                                                                                  
>>> referenced by UnitTest_PushFinalizeHook.cpp                                                                                                                                                                    
>>>               CMakeFiles/KokkosCore_UnitTest_PushFinalizeHook.dir/UnitTest_PushFinalizeHook.cpp.o:(__hip_register_globals)                 
>>> referenced by UnitTest_PushFinalizeHook.cpp                                                                                                                                                                    
>>>               CMakeFiles/KokkosCore_UnitTest_PushFinalizeHook.dir/UnitTest_PushFinalizeHook.cpp.o:(__hip_register_globals)                                                                                     
>>> referenced by UnitTest_PushFinalizeHook.cpp                                                                                                                                                                    >>>               CMakeFiles/KokkosCore_UnitTest_PushFinalizeHook.dir/UnitTest_PushFinalizeHook.cpp.o:(__hip_register_globals)                 
>>> referenced 42 more times                                                                                                                                                                                       
                                                                                                                                                                                                                   
ld.lld: error: undefined symbol: __hipRegisterFatBinary                                                                                                                                                            
>>> referenced by UnitTest_PushFinalizeHook.cpp                                                                                                                                                                    
>>>               CMakeFiles/KokkosCore_UnitTest_PushFinalizeHook.dir/UnitTest_PushFinalizeHook.cpp.o:(__hip_module_ctor)                                                                                          
>>> referenced by Kokkos_Core.cpp                                                                                                                                                                                  >>>               Kokkos_Core.cpp.o:(__hip_module_ctor) in archive ../src/libkokkoscore.a                                                                                                                          
>>> referenced by Kokkos_HIP_Instance.cpp                                                                                                                                                                          
>>>               Kokkos_HIP_Instance.cpp.o:(__hip_module_ctor) in archive ../src/libkokkoscore.a                                                                                                                  >>> referenced 12 more times                                                                                                                                                                                       
                                                                                                                                                                                                                   
ld.lld: error: undefined symbol: __hipUnregisterFatBinary                                                                                                                                                          
>>> referenced by UnitTest_PushFinalizeHook.cpp                                                          
>>>               CMakeFiles/KokkosCore_UnitTest_PushFinalizeHook.dir/UnitTest_PushFinalizeHook.cpp.o:(__hip_module_dtor)                                                                                          
>>> referenced by Kokkos_Core.cpp                                                                        
>>>               Kokkos_Core.cpp.o:(__hip_module_dtor) in archive ../src/libkokkoscore.a                                                                                                                          
>>> referenced by Kokkos_HIP_Instance.cpp                                                                                                                                                                          
>>>               Kokkos_HIP_Instance.cpp.o:(__hip_module_dtor) in archive ../src/libkokkoscore.a                                                                                                                  
>>> referenced 12 more times                                                                                                                                                                                       
                                                                                                                                                                                                                   
ld.lld: error: undefined symbol: hipSetDevice
>>> referenced by Kokkos_HIP_Instance.cpp:236 (/ccs/home/turcksin/crusher/kokkos/core/src/HIP/Kokkos_HIP_Instance.cpp:236)
>>>               Kokkos_HIP_Instance.cpp.o:(Kokkos::Impl::HIPInternal::initialize(int, ihipStream_t*, bool)) in archive ../src/libkokkoscore.a

ld.lld: error: undefined symbol: hipEventCreate
>>> referenced by Kokkos_HIP_Instance.cpp:300 (/ccs/home/turcksin/crusher/kokkos/core/src/HIP/Kokkos_HIP_Instance.cpp:300)
>>>               Kokkos_HIP_Instance.cpp.o:(Kokkos::Impl::HIPInternal::initialize(int, ihipStream_t*, bool)) in archive ../src/libkokkoscore.a

ld.lld: error: undefined symbol: hipMemset
>>> referenced by Kokkos_HIP_Instance.cpp:305 (/ccs/home/turcksin/crusher/kokkos/core/src/HIP/Kokkos_HIP_Instance.cpp:305)
>>>               Kokkos_HIP_Instance.cpp.o:(Kokkos::Impl::HIPInternal::initialize(int, ihipStream_t*, bool)) in archive ../src/libkokkoscore.a
>>> referenced by Kokkos_HIP_Instance.cpp:352 (/ccs/home/turcksin/crusher/kokkos/core/src/HIP/Kokkos_HIP_Instance.cpp:352)
>>>               Kokkos_HIP_Instance.cpp.o:(Kokkos::Impl::HIPInternal::scratch_flags(unsigned long)) in archive ../src/libkokkoscore.a

```

**Comment by jjellio** [comment]
Are yous setting `--offload-arch=gfx90a --rocm-path=/opt/rocm-5.2.0 -x hip ` ? and the proper link flag: `-x none`

These are the relevant bits of the CMake I use w/Trilinos

```
 "-DKokkos_ARCH_ZEN3:BOOL=ON" \
 "-DKokkos_ENABLE_SERIAL:BOOL=ON" \
 "-DTpetra_INST_SERIAL:BOOL=ON" \
 "-DKokkos_ENABLE_OPENMP:BOOL=OFF" \
 "-DTpetra_INST_OPENMP:BOOL=OFF" \
 "-DTrilinos_ENABLE_OpenMP:BOOL=OFF" \
 "-DKokkos_ENABLE_HIP:BOOL=ON" \
 "-DTpetra_INST_HIP:BOOL=ON" \
 "-DKokkos_ARCH_VEGA90A:BOOL=ON" \
 "-DCMAKE_EXE_LINKER_FLAGS=-x none " \
 "-DCMAKE_SHARED_LINKER_FLAGS=-x none" \
 "-DCMAKE_LINKER=/opt/rocm-5.2.0/bin/amdclang++" \
 "-DCMAKE_CXX_LINK_EXECUTABLE=<CMAKE_LINKER> <FLAGS> <CMAKE_CXX_LINK_FLAGS> -x none <LINK_FLAGS> <OBJECTS> -o <TARGET> <LINK_LIBRARIES>" \
 "-DCMAKE_CXX_CREATE_SHARED_LIBRARY=<CMAKE_LINKER> <CMAKE_SHARED_LIBRARY_CXX_FLAGS> <LANGUAGE_COMPILE_FLAGS> -x none <LINK_FLAGS> <CMAKE_SHARED_LIBRARY_CREATE_CXX_FLAGS> <SONAME_FLAG><TARGET_SONAME> -o <TARGET> <OBJECTS> <LINK_LIBRARIES>" \
 "-DCMAKE_Fortran_LINK_EXECUTABLE=<CMAKE_LINKER> <FLAGS> <CMAKE_CXX_LINK_FLAGS> -x none <LINK_FLAGS> <OBJECTS> -o <TARGET> <LINK_LIBRARIES>" \
 "-DCMAKE_Fortran_CREATE_SHARED_LIBRARY=<CMAKE_Fortran_COMPILER> <CMAKE_SHARED_LIBRARY_Fortran_FLAGS> <LANGUAGE_COMPILE_FLAGS> <LINK_FLAGS> <CMAKE_SHARED_LIBRARY_CREATE_Fortran_FLAGS> <SONAME_FLAG><TARGET_SONAME> -o <TARGET> <OBJECTS> <LINK_LIBRARIES>" \
 "-DCMAKE_CXX_COMPILER=/opt/rocm-5.2.0/bin/amdclang++" \
 "-DCMAKE_C_COMPILER=/opt/rocm-5.2.0/bin/amdclang" \
 "-DCMAKE_Fortran_COMPILER=/opt/rocm-5.2.0/bin/amdflang" \
 "-DCMAKE_CXX_FLAGS=--offload-arch=gfx90a --rocm-path=/opt/rocm-5.2.0 -x hip -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -I/opt/cray/pe/mpich/8.1.18/ofi/crayclang/10.0/include -I/opt/rocm-5.2.0/include -g " \
 "-DCMAKE_CXX_STANDARD=17" \
```

**Comment by Rombur** [comment]
What's the point on setting `ROCM_DEFAULT=OFF`, if I have to set the flags by hand myself. You just want a way to override the flags that are chosen automatically?

**Comment by jjellio** [comment]
Because I don't want to use hipcc - (I build through Cray as well as AMD)

**Comment by Rombur** [comment]
I get that but we set `ROCM_DEFAULT` to let CMake know that it needs to add the flags for HIP itself. I just tried `amdclang` and, with the current code`, the flags are set correctly.

**Comment by jjellio** [comment]
This blocks being able to use CXX=CC (Cray PE wrappers) regardless of AMD/Cray. I also build directly w/clang (AMD or Cray) - this is essential to setting non standard flags that HIPCC chokes on.

I set all the flags anyway

**Comment by jjellio** [comment]
@Rombur - How is this intended to work?  I think what you are saying is,
1. set amdclang++ as CXX (but don't specify -x hip/none or ROCM_PATH)
2. Cmake should then "find" ROCM

I'm happy to try this - my builds predate CMake supporting this stuff, so if we move in that direction, I'd *like* to be able to **NOT** specify all the flags I currently do.

**Comment by jjellio** [comment]
Here's the familiar error:

```
-- Architectures:
--  ZEN3
--  VEGA90A
CMake Error at packages/kokkos/cmake/kokkos_tpls.cmake:29 (MESSAGE):
  Enabled TPL ROCM inside TriBITS build, but this can only be enabled in a
  standalone build
Call Stack (most recent call first):
  packages/kokkos/cmake/kokkos_tpls.cmake:47 (KOKKOS_TPL_OPTION)
  packages/kokkos/cmake/kokkos_tribits.cmake:247 (INCLUDE)
  packages/kokkos/CMakeLists.txt:188 (KOKKOS_SETUP_BUILD_ENVIRONMENT)
```

**Comment by Rombur** [comment]
I see. When I build Kokkos alone, I only have to change the compiler from `hipcc` to `amdclang++` and everything works.  I don't know why the behavior is different when using Tribits. I'll look into it.

**Comment by jjellio** [comment]
It's generally been because Kokkos/Trilinos Cmake isn't 100% compatible.

Before you invest too much time, let's see if Ross has some comments (I was hoping his big TRIBITS update was going to include letting Kokkos build HIP more naturally)

@bartlettroscoe

**Comment by bartlettroscoe** [comment]
> Before you invest too much time, let's see if Ross has some comments (I was hoping his big TRIBITS update was going to include letting Kokkos build HIP more naturally)

The current version of TriBITS in Trilinos should be propagating all usage requirements from Kokkos targets to downstream targets (both inside and outside of Trilinos).  Does the Kokkos TriBITS build set compiler and linker options using `target_compile_options()` and `target_link_options()`, for example?

**Comment by jjellio** [comment]
Ross, I don't know the details of Kokkos / Trilinos CMake stuff - I do know in the past, there has been extra work needed.

Perhaps, this TRIBITS error is simply a relic of the past and can now be deleted? I didn't dig too deep to try to disable it (because when I tried forcing the variable that caused this to arise, it triggered other problems later (because Kokkos stopped thinking it was being built inside Trilinos).

I can push on it a little more though. I'd love to see this functionality work (using native compilers w/Trilinos) - We'd then see if the new TRIBITS update actually propagates all the stuff Kokkos adds (HIP, all it's libs, and the compiler/linker settings) - As Bruno points out, I am manually specifying all that currently (and this causes some headaches downstream, when an app uses Trilinos - we need to make sure to inherit all of the compiler, linker and flags exactly - or things break).  With ATDM EMPIRE, I also need to set CMake's linker variables for SHARED and EXE's, because Trilinos doesn't propagate those CMake variables into Trilinos.cmake. (not Trilinos_SHARED_LIBRARY_FLAGS or whatever)

**Comment by bartlettroscoe** [comment]
> I'd love to see this functionality work (using native compilers w/Trilinos)

@jjellio, you can't get this to work through the MPI compiler wrappers?

> We'd then see if the new TRIBITS update actually propagates all the stuff Kokkos adds (HIP, all it's libs, and the compiler/linker settings) 

If Kokkos is adding all of those to the Kokkos targets usage requirements, they should propagate downstream.  I would have to look to see what Kokkos is doing in its TriBITS build vs. its native CMake build.

**Comment by jjellio** [comment]
Ross, I have it working.  The problem is, I have to manually edit`kokkos/cmake/kokkos_tpls.cmake#L42` to set disable Kokkos/Cmake from treating ROCM as a TPL.  Ideally, we should be able to build HIP with AMD compiler directly (and Kokkos can - and if you do my hack, Trilinos can as well) -- This is essential for testing non-AMD compilers (e.g., Cray) - and it is required if you need to change the default AMD optimization flags (currently, hipcc sets 2 gpu optimizations, that cannot be disabled w/out hacking hipcc - it barfs if you set duplicate flags to try and override). Or, you compiler with amdclang directly - and you can choose whatever flags you want. (or use Cray).  These flags specifically impact working around register allocation bugs, so are pretty useful to control.

TLDR: the issue requires editing Kokkos' CMake  as-is, if this is just a relic of a past TRIBITS+Kokkos+Trilinos requirement, perhaps nowadays, it can go away??

**Comment by Rombur** [comment]
I have talked to @dalg24 and we are pretty sure that the problem is in Kokkos. I problably won't get it done by the time we branch Kokkos 4.0 but I will let you know when it works @jjellio

**Comment by bartlettroscoe** [comment]
> The problem is, I have to manually editkokkos/cmake/kokkos_tpls.cmake#L42 to set disable Kokkos/Cmake from treating ROCM as a TPL.

@jjellio, as a workaround for now so you don't have to manually edit the line:

https://github.com/kokkos/kokkos/blob/61d7db55fceac3318c987a291f77b844fd94c165/cmake/kokkos_tpls.cmake#L47

could you not just submit a PR to Kokkos to change that block of code as:

```
if ("${Kokkos_ROCM_DEFAULT}" STREQUAL "")
  if (KOKKOS_ENABLE_HIP AND NOT KOKKOS_CXX_COMPILER_ID STREQUAL HIPCC)
    set(Kokkos_ROCM_DEFAULT ON)
  else()
    set(Kokkos_ROCM_DEFAULT OFF)
  endif()
endif()
kokkos_tpl_option(ROCM    ${Kokkos_ROCM_DEFAULT})
```

?

That way, if you set `-DKokkos_ROCM_DEFAULT=OFF`, then the default logic would not run.  (Note, it is a good idea to namespace vars that can be overridden in the cache like this.)

**Comment by jjellio** [comment]
Thanks Ross!

**Comment by Rombur** [comment]
@jjellio We have merged a partial fix to this issue. When using Trilinos with `amdclang`, Kokkos won't error out but you need to provide the flags yourself. If you don't give any flags the configuration will work but it won't compile and the error message is not very helpful.

**Comment by dalg24** [comment]
It also means you will need to repeat the flags downstream

**Comment by jjellio** [comment]
Thanks!  This should be a big help.  I'll see about giving it a spin the next time im on the LLNL machine

**Comment by jjellio** [comment]
This has been fixed for a while ... I'll close it, unless there is something else?
